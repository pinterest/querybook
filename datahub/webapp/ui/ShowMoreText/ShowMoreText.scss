.ShowMoreText {
    .ShowMoreText-click {
        cursor: pointer;
        white-space: nowrap;
        text-decoration: underline;
    }
}
