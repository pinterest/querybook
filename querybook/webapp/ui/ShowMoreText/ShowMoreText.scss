.ShowMoreText {
    .ShowMoreText-click {
        cursor: pointer;
        white-space: nowrap;
        text-decoration: underline;
        display: inline;
        margin-left: 4px;
    }
    &.is-list {
        span {
            margin-left: 0px;
            display: block;
        }
    }
}
