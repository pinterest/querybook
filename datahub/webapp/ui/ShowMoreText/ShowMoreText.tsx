import React, { useMemo } from 'react';

import './ShowMoreText.scss';

interface IShowMoreTextProps {
    text: string;
    length?: number;
    seeLess?: boolean;
    className?: string;
}

export const ShowMoreText: React.FunctionComponent<IShowMoreTextProps> = ({
    text,
    length = 100,
    seeLess = false,
    className = '',
}) => {
    text = text || '';
    const combinedClassName = useMemo(() => `ShowMoreText ${className}`, [
        className,
    ]);

    const [expanded, setExpanded] = React.useState(false);
    const toggleSeeMoreClick = (e: React.SyntheticEvent) => {
        if (e) {
            e.stopPropagation();
            e.preventDefault();
        }
        setExpanded(!expanded);
    };

    if (text.length === 0) {
        return null;
    } else if (text.length >= length) {
        // exceeding length requirement
        if (!expanded) {
            return (
                <span className={combinedClassName}>
                    {text.slice(0, length)}{' '}
                    <a
                        className="ShowMoreText-click"
                        onClick={toggleSeeMoreClick}
                    >
                        show more
                    </a>
                </span>
            );
        } else {
            const seeLessSection = seeLess ? (
                <>
                    {' '}
                    <a
                        className="ShowMoreText-click"
                        onClick={toggleSeeMoreClick}
                    >
                        show less
                    </a>
                </>
            ) : null;

            return (
                <span className={combinedClassName}>
                    {text}
                    {seeLessSection}
                </span>
            );
        }
    }

    // normal case, text within the number of chars
    return <span className={combinedClassName}>{text}</span>;
};
