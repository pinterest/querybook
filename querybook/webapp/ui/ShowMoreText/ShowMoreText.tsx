import React from 'react';
import clsx from 'clsx';

import './ShowMoreText.scss';

interface IShowMoreTextProps {
    text: string | string[];
    length?: number;
    seeLess?: boolean;
    className?: string;
}

export const ShowMoreText: React.FunctionComponent<IShowMoreTextProps> = ({
    text,
    length = 100,
    seeLess = false,
    className = '',
}) => {
    text = text ?? '';

    const [expanded, setExpanded] = React.useState(false);
    const isList = Array.isArray(text);
    const max = length ?? (isList ? 4 : 100);

    const toggleSeeMoreClick = (e: React.SyntheticEvent) => {
        if (e) {
            e.stopPropagation();
            e.preventDefault();
        }
        setExpanded(!expanded);
    };

    const combinedClassName = clsx({
        ShowMoreText: true,
        [className]: className,
        'is-list': isList,
    });

    if (text.length === 0) {
        return null;
    } else if (text.length >= max && !expanded) {
        // exceeding length requirement
        const truncatedText = text.slice(0, max);
        return (
            <span className={combinedClassName}>
                {isList
                    ? (truncatedText as string[]).map((line, idx) => (
                          <span key={idx}>{line}</span>
                      ))
                    : truncatedText}
                <span
                    className="ShowMoreText-click"
                    onClick={toggleSeeMoreClick}
                >
                    show more
                </span>
            </span>
        );
    } else {
        // normal case, text within the max
        const seeLessSection =
            seeLess && text.length >= max ? (
                <span
                    className="ShowMoreText-click"
                    onClick={toggleSeeMoreClick}
                >
                    show less
                </span>
            ) : null;

        return (
            <span className={combinedClassName}>
                {isList
                    ? (text as string[]).map((line, idx) => (
                          <span key={idx}>{line}</span>
                      ))
                    : text}
                {seeLessSection}
            </span>
        );
    }
};
