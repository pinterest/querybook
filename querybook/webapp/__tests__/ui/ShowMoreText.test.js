import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { ShowMoreText } from '../../ui/ShowMoreText/ShowMoreText';

it('renders without crashing', () => {
    shallow(<ShowMoreText text="test" length={2} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<ShowMoreText text="test" length={2} />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});
