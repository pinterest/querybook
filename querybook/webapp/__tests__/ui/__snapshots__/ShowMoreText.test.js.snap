// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<span
  className="ShowMoreText"
>
  te
  <span
    className="ShowMoreText-click"
    onClick={[Function]}
  >
    show more
  </span>
</span>
`;
